src/mon/OSDMonitor.cc

```diff
@@ -606,6 +762,104 @@ void OSDMonitor::update_from_paxos(bool *need_bootstrap)
   }
 }
 
+int OSDMonitor::register_cache_with_pcm()
+{
+  if (mon_memory_target <= 0 || mon_memory_min <= 0) {
+    derr << __func__ << " Invalid memory size specified for mon caches."
+         << " Caches will not be auto-tuned."
+         << dendl;
+    return -EINVAL;
+  }
+  uint64_t base = mon_memory_base;
+  double fragmentation = mon_memory_fragmentation;
+  // For calculating total target memory, consider rocksdb cache size.
+  uint64_t target = mon_memory_target;
+  uint64_t min = mon_memory_min;
+  uint64_t max = min;
+
+  // Apply the same logic as in bluestore to set the max amount
+  // of memory to use for cache. Assume base memory for OSDMaps
+  // and then add in some overhead for fragmentation.
+  uint64_t ltarget = (1.0 - fragmentation) * target;
+  if (ltarget > base + min) {
+    max = ltarget - base;
+  }
+
+  rocksdb_binned_kv_cache = mon->store->get_priority_cache();
+  ceph_assert(rocksdb_binned_kv_cache);
+
+  int r = _set_cache_ratios();
+  if (r < 0) {
+    derr << __func__ << " Cache ratios for pcm could not be set."
+         << " Review the kv (rocksdb) and mon_memory_target sizes."
+         << dendl;
+    return -EINVAL;
+  }
+
+  pcm = std::make_shared<PriorityCache::Manager>(
+      cct, min, max, target);
+  pcm->insert("kv", rocksdb_binned_kv_cache);
+  pcm->insert("inc", inc_cache);
+  pcm->insert("full", full_cache);
+  dout(10) << __func__ << " pcm target: " << target
+           << " pcm max: " << max
+           << " pcm min: " << min
+           << " inc_osd_cache size: " << inc_osd_cache.get_size()
+           << dendl;
+  return 0;
+}
+
+int OSDMonitor::_set_cache_ratios()
+{
+  double old_cache_kv_ratio = cache_kv_ratio;
+
+  // Set the cache ratios for kv(rocksdb), inc and full caches
+  cache_kv_ratio = (double)rocksdb_cache_size / (double)mon_memory_target;
+  if (cache_kv_ratio >= 1.0) {
+    derr << __func__ << " Cache kv ratio (" << cache_kv_ratio
+         << ") must be in range [0,<1.0]."
+         << dendl;
+    cache_kv_ratio = old_cache_kv_ratio;
+    return -EINVAL;
+  }
+  rocksdb_binned_kv_cache->set_cache_ratio(cache_kv_ratio);
+  cache_inc_ratio = cache_full_ratio = (1.0 - cache_kv_ratio) / 2;
+  inc_cache->set_cache_ratio(cache_inc_ratio);
+  full_cache->set_cache_ratio(cache_full_ratio);
+
+  dout(10) << __func__ << " kv ratio " << cache_kv_ratio
+           << " inc ratio " << cache_inc_ratio
+           << " full ratio " << cache_full_ratio
+           << dendl;
+  return 0;
+}
+
+uint64_t OSDMonitor::calculate_used_cache_bytes(bool cache_type_full)
+{
+  uint64_t total_bytes = 0;
+  uint64_t features = mon->get_quorum_con_features();
+  epoch_t first = get_first_committed();
+  epoch_t latest = osdmap.get_epoch();
+
+  for (epoch_t e = latest; e >= first && e > 0; e--) {
```

Comment:
this approach has a few issues:
- in a steady state, the cache will always be full, so it will recalculate this way with every new map
- it uses the cache itself when reading, and may read versions off disk that aren't useful to cache, and may change the cache size
- it's iterating over potentially 1000s of maps

A less expensive approach would be modifying SimpleLRU to keep track of the number of bytes it contains directly.
Since SimpleLRU is only used in three places, the simplest approach might be assuming a length() method exists, and adding one that returns sizeof(_Header) to os/filestore/DBObjectMap.h's _Header class. The other two users store bufferlists, which already has a length() method.

Then in SimpleLRU you can increment/decrement the bytes as appropriate when things are added or removed from the cache, and querying the size requires no recalculation.

Reply:
Done as per your suggestion. 

I have also added a divide-by-zero guard in places where bytes per osdmap calculations are done. I don't think these are really required, but since I noticed a weird behavior while testing on my laptop, I decided to add them just to be safe. I observed that the PCM returned odd  values for the last cache that was inserted (full cache) against what was allocated. I think it could be because of not enough memory on my machine.